### PR TITLE
Fix (non-reduction) ops over a dimension for n-dimensional empty tens…

### PR DESCRIPTION
**Pull request by gchanan**
…ors (CPU).

This includes (mainly) CPU fixes; CUDA fixes are a little more involved because you can't use an empty grid.
This also includes a fix for index_copy, which checked that self.size(dim) == src.size(0), which isn't correct (the same dimension should be compared).
Finally, also includes a fix for CUDA flip (although it's not tested yet), to get the stride using multiplication rather than division to avoid divide-by-0.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -6106,6 +6106,116 @@ def test_tensor_shape_empty(self):
...
+        devices = ['cpu']  # if not torch.cuda.is_available() else ['cpu', 'cuda']
+        for device in devices:
+            shape = (0, 1, 2, 0)
+            x = torch.randn(shape, device=device)
+
+            # size stride
+            self.assertEqual(0, x.size(3))
+            self.assertEqual(2, x.size(2))
+            self.assertEqual(2, x.stride(0))
+            self.assertEqual(1, x.stride(2))
```
Comment (ezyang) [inline test/test_torch.py]: Did we document the "stride for zero-size dims" invariant anywhere? If we're going to test that the stride matches exactly, that implies that the stride we select in these situations is part of the public API.

File: test/test_torch.py [hunk 2]
```diff
@@ -6106,6 +6106,116 @@ def test_tensor_shape_empty(self):
...
+            self.assertEqual(x, x.flip(2))
+
+            # unbind
+            self.assertEqual((), x.unbind(0))
+            self.assertEqual((torch.empty((0, 1, 0), device=device), torch.empty((0, 1, 0), device=device)),
+                             x.unbind(2))
+
+            # cross
+            y = torch.randn((0, 1, 3, 0), device=device)
+            self.assertEqual(y.shape, torch.cross(y, y).shape)
```
Comment (ezyang) [inline test/test_torch.py]: I guess we don't really need a dim= test here

File: test/test_torch.py (hunk 1 above)
Comment (gchanan) [reply test/test_torch.py]: Good question, and I'm okay changing this to just check that you get back some reasonable value if we want to do that.  But code like flip_cuda (see this PR) already assumes there is some canonical "stride for contiguous tensors" (and yes, numel() == 0 tensors are considered contiguous).  So this isn't really a departure from that.  Also, we do the same thing as NumPy, and NumPy doesn't document their numel() == 0 strides as far as I can tell.

File: test/test_torch.py [hunk 3]
```diff
@@ -6106,6 +6106,116 @@ def test_tensor_shape_empty(self):
...
+
+            # renorm
+            self.assertEqual(shape, torch.renorm(x, 1, 0, 5).shape)
```
Comment (ezyang) [inline test/test_torch.py]: It'd be nice if the tests explicitly said which of these were pointwise, and which ones had more complicated shape equations.

File: test/test_torch.py [hunk 4]
```diff
@@ -6106,6 +6106,116 @@ def test_tensor_shape_empty(self):
...
+            self.assertEqual(shape, torch.renorm(x, 1, 2, 5).shape)
+
+            # sort
+            self.assertEqual([shape, shape], [z.shape for z in torch.sort(x, dim=0)])
```
Comment (ezyang) [inline test/test_torch.py]: (values, indices) right? A little oblique, but OK.

File: test/test_torch.py (hunk 2 above)
Comment (gchanan) [reply test/test_torch.py]: It picks the first one of size 3.

File: aten/src/ATen/native/Indexing.cpp [hunk 5]
```diff
@@ -309,7 +305,7 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
   }
   auto sourceSlicedSizes = std::vector<int64_t>(source.sizes());
   if (sourceSlicedSizes.size() > 0) {
-    sourceSlicedSizes.erase(sourceSlicedSizes.begin());
+    sourceSlicedSizes.erase(sourceSlicedSizes.begin() + dim);
```
Comment (ezyang) [inline aten/src/ATen/native/Indexing.cpp]: Which test is for this fix? (For my education.)

File: test/test_torch.py (hunk 3 above)
Comment (gchanan) [reply test/test_torch.py]: I thought about doing this sort of thing, but I'm not convinced it's a good idea.  In general I'd have to document the shape equation (it's already in the docs), and they would probably get out of date soon.

File: aten/src/ATen/native/Indexing.cpp (hunk 5 above)
Comment (gchanan) [reply aten/src/ATen/native/Indexing.cpp]: I'm not sure exactly what the question is -- are you asking which new test case corresponds to this?  In that case, it's any index_copy that doesn't use dim 0.

File: aten/src/ATen/native/cuda/TensorTransformations.cu [hunk 6]
```diff
@@ -99,18 +100,22 @@ Tensor flip_cuda(const Tensor& self, IntList dims) {
...
-  // stride_contiguous is the stride of non-contiguous tensor after called contiguous(), it is used to compute indices for each element in non-contiguous tensor
+  // stride_contiguous is the stride of non-contiguous tensor after calling contiguous(),
+  // it is used to compute indices for each element in non-contiguous tensor
   Tensor stride_contiguous = at::zeros({total_dims}, kLong);
   int64_t* stride_contiguous_d = stride_contiguous.data<int64_t>();
-  int64_t tmp = N;
-  for (int64_t i = 0; i < total_dims; i++) {
-    tmp = tmp / shape[i];
-    stride_contiguous_d[i] = tmp;
+  for (int64_t i = total_dims - 1; i >=0; i--) {
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/TensorTransformations.cu]: nit: space after >=

**Review by ezyang — APPROVED** [review]
I didn't do a close enough review to say that I am confident that I caught bugs, but I think this is good enough to go in for now.

**Pull request closed without merging**